### PR TITLE
Number porting

### PR DESCRIPTION
**Pull request by VladyslavBanul**
https://app.asana.com/0/1201128813141482/1201932298109365/f
https://app.asana.com/0/1201128813141482/1201275099482093/f
https://app.asana.com/0/1201128813141482/1201275099482097/f
https://app.asana.com/0/1201128813141482/1201275099482095/f
https://app.asana.com/0/1201128813141482/1201275099482091/f
https://app.asana.com/0/1201128813141482/1201275099482099/f

### CODE REVIEW

File: src/levels/subscription/components/SubscriptionNumbersPage/TabsContent/PortingRequests/Modals/RequestInfoModal/Tabs/TabsContent/EnterpriseInfo.tsx [hunk 1]
```diff
@@ -0,0 +1,12 @@
+//NOTE: it's unfinished component that will be finished in future
+import { observer } from "mobx-react-lite";
+
+import { requestInfoStyles } from "../../../../styles";
```
Comment (SergeyKoliadintsev) [inline src/levels/subscription/components/SubscriptionNumbersPage/TabsContent/PortingRequests/Modals/RequestInfoModal/Tabs/TabsContent/EnterpriseInfo.tsx:4]: change import please !!!

File: src/levels/subscription/components/SubscriptionNumbersPage/TabsContent/PortingRequests/Modals/RequestInfoModal/Tabs/TabsContent/PortingNumbers.tsx [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+  );
+
+  return numbersRanges.length ? (
+    <Table
+      title={t("Porting numbers")}
+      columns={columns}
+      data={numbersRanges}
+    />
+  ) : (
+    <div>there are no number ranges</div>
```
Comment (SergeyKoliadintsev) [inline src/levels/subscription/components/SubscriptionNumbersPage/TabsContent/PortingRequests/Modals/RequestInfoModal/Tabs/TabsContent/PortingNumbers.tsx:67]: Add translate please

File: src/levels/subscription/components/SubscriptionNumbersPage/TabsContent/PortingRequests/Modals/RequestInfoModal/Tabs/TabsContent/PortingRequestDetails.tsx [hunk 3]
```diff
@@ -0,0 +1,100 @@
+import { useEffect, useMemo, useState } from "react";
+import { observer } from "mobx-react-lite";
+import { useTranslation } from "react-i18next";
+
+import PortingRequestsStore from "storage/singletons/PortingRequests";
+
+import { requestInfoStyles } from "../../../../styles";
```
Comment (SergeyKoliadintsev) [inline src/levels/subscription/components/SubscriptionNumbersPage/TabsContent/PortingRequests/Modals/RequestInfoModal/Tabs/TabsContent/PortingRequestDetails.tsx:7]: change import please !!!

File: src/levels/subscription/components/SubscriptionNumbersPage/TabsContent/PortingRequests/Modals/RequestInfoModal/Tabs/TabsContent/PortingRequestDetails.tsx [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+  const { t } = useTranslation();
+  const classes = requestInfoStyles();
+
+  const {
+    currentPortingRequest: {
+      instance: { createdOn, updatedOn },
+      request: {
+        portId,
+        status,
+        // dueDate,
```
Comment (SergeyKoliadintsev) [inline src/levels/subscription/components/SubscriptionNumbersPage/TabsContent/PortingRequests/Modals/RequestInfoModal/Tabs/TabsContent/PortingRequestDetails.tsx:19]: you need this comment ?

File: src/levels/subscription/components/SubscriptionNumbersPage/TabsContent/PortingRequests/Modals/RequestInfoModal/index.tsx [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+import Modal from "components/Modal";
+import RouteIndependedTabs from "components/Tabs/RouteIndependedTabs";
+import PortingRequestDetails from "./Tabs/TabsContent/PortingRequestDetails";
+import PortingNumbers from "./Tabs/TabsContent/PortingNumbers";
+import { Cross, InfoIcon } from "components/Icons";
+import Tooltip from "components/Tooltip";
+import ButtonWithIcon from "components/common/Form/ButtonWithIcon";
+import Documents from "./Tabs/TabsContent/Documents";
+
+import { requestModalStyles } from "../../styles";
```
Comment (SergeyKoliadintsev) [inline src/levels/subscription/components/SubscriptionNumbersPage/TabsContent/PortingRequests/Modals/RequestInfoModal/index.tsx:22]: change import please !!!

File: src/levels/subscription/components/SubscriptionNumbersPage/TabsContent/PortingRequests/index.tsx [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+import { getIsLoading } from "utils/functions/getIsLoading";
+
+import { Plus } from "components/Icons";
+import Table from "components/Table";
+import TableSkeleton from "components/Table/Skeleton";
+import ButtonWithIcon from "components/common/Form/ButtonWithIcon";
+import RequestInfoModal from "./Modals/RequestInfoModal";
+import AddPortingRequestModal from "./Modals/AddPortingRequestModal";
+
+import styles from "./styles";
```
Comment (SergeyKoliadintsev) [inline src/levels/subscription/components/SubscriptionNumbersPage/TabsContent/PortingRequests/index.tsx:19]: Perfect variant ! Please remember it !

File: src/levels/subscription/components/SubscriptionNumbersPage/TabsContent/PortingRequests/styles.ts [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+  },
+  modalLink: {
+    cursor: "pointer",
+    color: theme.palette.primary.newLink,
+    display: "contents",
+  },
+  noNumbersBlock: {
+    display: "flex",
+    margin: "auto",
+    gap: 20,
```
Comment (SergeyKoliadintsev) [inline src/levels/subscription/components/SubscriptionNumbersPage/TabsContent/PortingRequests/styles.ts:19]: add theme spacing please

File: src/levels/subscription/components/SubscriptionNumbersPage/TabsContent/PortingRequests/styles.ts [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+    color: theme.palette.primary.white,
+    border: "none",
+  },
+  cancelButton: {
+    background: theme.palette.status.error,
+    color: theme.palette.primary.white,
+    border: "none",
+
+    "&>* svg": {
+      fill: theme.palette.primary.white + "!important",
```
Comment (SergeyKoliadintsev) [inline src/levels/subscription/components/SubscriptionNumbersPage/TabsContent/PortingRequests/styles.ts:48]: change please this to `          ${}          `

File: src/levels/subscription/components/SubscriptionNumbersPage/TabsContent/PortingRequests/styles.ts [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+    },
+  },
+  buttonsWrapper: {
+    position: "fixed",
+    display: "flex",
+    gap: 10,
```
Comment (SergeyKoliadintsev) [inline src/levels/subscription/components/SubscriptionNumbersPage/TabsContent/PortingRequests/styles.ts:54]: add theme spacing please

File: src/levels/subscription/components/SubscriptionNumbersPage/TabsContent/PortingRequests/styles.ts [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+    borderRadius: 6,
+    fontSize: "1.4rem",
+    lineHeight: "1.6rem",
+    color: theme.palette.secondary.text,
+    padding: "10px 20px",
+  },
+  numbersFlex: {
+    display: "flex",
+    flexWrap: "wrap",
+    gap: 20,
```
Comment (SergeyKoliadintsev) [inline src/levels/subscription/components/SubscriptionNumbersPage/TabsContent/PortingRequests/styles.ts:147]: ` add   `  please `theme speacing` this

File: src/levels/subscription/components/SubscriptionNumbersPage/TabsContent/PortingRequests/styles.ts [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+  tooltipIcon: {
+    fill: theme.palette.status.error,
+    height: 18,
+    width: 18,
+  },
+  successStatus: { color: theme.palette.status.success },
+  failStatus: {
+    display: "flex",
+    alignItems: "center",
+    gap: 5,
```
Comment (SergeyKoliadintsev) [inline src/levels/subscription/components/SubscriptionNumbersPage/TabsContent/PortingRequests/styles.ts:166]: ` add   `  please `theme speacing` this

File: src/levels/subscription/components/SubscriptionNumbersPage/TabsContent/PortingRequests/styles.ts [hunk 12]
```diff
@@ -0,0 +1,177 @@
...
+    color: theme.palette.status.error,
+  },
+}));
+
+export const documentsStyles = makeStyles(() => ({
+  cardsWrapper: {
+    display: "flex",
+    flexDirection: "column",
+    gap: 20,
```
Comment (SergeyKoliadintsev) [inline src/levels/subscription/components/SubscriptionNumbersPage/TabsContent/PortingRequests/styles.ts:175]: ` add   `  please `theme speacing` this

File: src/storage/singletons/PortingRequests/index.tsx [hunk 13]
```diff
@@ -0,0 +1,241 @@
+import { makeAutoObservable, runInAction } from "mobx";
+import { AxiosResponse } from "axios";
+
+import configStore from "../Config";
+import { request } from "services/api";
+import { errorNotification } from "utils/functions/notifications";
+import { DocumentsType, PortingRequirements } from "utils/types/numbers";
+import TablePagination from "../TablePagination";
+
+type RequestType = {
```
Comment (SergeyKoliadintsev) [inline src/storage/singletons/PortingRequests/index.tsx:10]: move this type in another file with types

File: src/storage/singletons/PortingRequests/index.tsx [hunk 14]
```diff
@@ -0,0 +1,241 @@
+import { makeAutoObservable, runInAction } from "mobx";
+import { AxiosResponse } from "axios";
+
```
Comment (SergeyKoliadintsev) [inline src/storage/singletons/PortingRequests/index.tsx:3]: Change file to TS

File: src/storage/singletons/PortingRequests/index.tsx [hunk 15]
```diff
@@ -0,0 +1,241 @@
...
+  numbers: Array<string>;
+  ranges: [
+    {
+      from: string;
+      to: string;
+    },
+  ];
+};
+
+type PortingRequestPayload = {
```
Comment (SergeyKoliadintsev) [inline src/storage/singletons/PortingRequests/index.tsx:30]: move this type in another file with types

File: src/utils/schemas/numbersPorting.ts [hunk 16]
```diff
@@ -0,0 +1,10 @@
+import { string, object } from "yup";
+import StringSchema from "yup/lib/string";
+
+export const countrySchema = () =>
+  object().shape({
+    countryCode: string().required("Please selelct a country"),
```
Comment (SergeyKoliadintsev) [inline src/utils/schemas/numbersPorting.ts:6]: Add translate please

**Comment by Coelmount** [comment]
@VladyslavBanul - issue 1: https://apio.copaco.netaxis.cloud/transactions/615520

I've fixed the upload attachment at the backend (note: still an issue with attachments larger than 1M)
However, the problem is that you just put the document content as such in the body.

You need to send it as a json object with the content in a key named "content" and with a "description" (string). As description you need to use the "id" of the document as found in the requirement API.

See also: https://docs.netaxis.solutions/draas/provisioning/api/08_numbers.html#add-attachments-to-a-porting-request

For example, for the required document named "loa", send:
{
   "description": "loa",
   "content": "<content>"
}

**Comment by Coelmount** [comment]
@VladyslavBanul  Secondly, this is not really a bug but something that was missing from API perspective as you requested but which I ignored (sorry for this). Now the API will return you "instance.errors" ==> it's an array of errors:
- if not errors, it's empty
- if there are errors, you will get 1 or more objects. Per error you can show a card. In the card print the created_on and the description.

image

**Comment by Coelmount** [comment]
@VladyslavBanul  - another item (not super important) is the structuring of the request details. Can you please group the different fields as defined in the requirements definition. Similar as how you structure the input form in the wizard in step 2

image

**Comment by Coelmount** [comment]
other problem is that the documents tab shows dummy data:
image


Upload doesn't work either of course

**Pull request merged**